### ISSUE TITLE
☂️ make initial load of canvaskit faster

### ISSUE THREAD

**Issue by franarolas** [flutter/flutter#76009]
Hello everyone!

This post is a request from the Flutter community to request an improvement on the start load time of Flutter Web.
As most of you know, the initial load time when using the canvas build is way too long for a production product, so most of the people are not choosing Flutter Web for their web apps.

Thank you all for voting this up, I hope that this gets solved as soon as possible.

Since the old post didn't make much sense in the context of what we are trying to achieve, I decided to rewrite the post.

<details>
<summary>Old post</summary>

> I'm experiencing a huge difference in the initial load time of the web app in the versions 1.26 and 1.27, the load time now is about 2-3 seconds slower.
> The lighthouse tool can't even test the app performance as it is too slow.
> Once the app is loaded, it is as fast as always.
> 
> It is affecting also the hot-reload time, it now takes about 10 seconds more to reload my project(A big project with more than 240 files and around 30000 written code lines).
> 
> I would be working with the 1.25 version(As everything was working fine, including SVGs), but I really need the Scrollbar implemented in 1.26.
> 
> ## Steps to Reproduce
> 
> Just create a new Flutter project and test the different versions initial load time(Without cache).
> 
> I'm not posting my Flutter doctor as it is unrelated.
> 
> ## Lighthouse audits in different versions:
> 
> Alpha - 1.27.0-1.0.pre:
> ![image](https://user-images.githubusercontent.com/42511706/107881705-dbda1200-6ee5-11eb-9074-7877d50038a7.png)
> 
> Beta - 1.26.0-17.5.pre:
> ![image](https://user-images.githubusercontent.com/42511706/107881714-e5fc1080-6ee5-11eb-82ef-de5b5db85b8f.png)
> 
> Beta - 1.25.0-8.3.pre:
> ![image](https://user-images.githubusercontent.com/42511706/107881741-f14f3c00-6ee5-11eb-9509-957eaddcf856.png)
> 
> Everything is tested in a production server, with the release version and the following pubspec:
> ```
> name: flutter_load_test
> description: A new Flutter project.
> 
> publish_to: 'none'
> 
> version: 1.0.0+1
> 
> environment:
>   sdk: ">=2.7.0 <3.0.0"
> 
> dependencies:
>   flutter:
>     sdk: flutter
> 
> dev_dependencies:
>   flutter_test:
>     sdk: flutter
> 
> flutter:
>   uses-material-design: false
> ```

</details>

**Comment by iapicca** [comment]
would be interesting to introduce a test based in [lighthouse](https://developers.google.com/web/tools/lighthouse) to prevent regressions

**Comment by azbuky** [comment]
@franarolas what web renderer are you using, html or canvaskit?

**Comment by franarolas** [comment]
> @franarolas what web renderer are you using, html or canvaskit?

@azbuky Always the default one, it should be always canvas as I am testing it through desktop browser

**Comment by TahaTesser** [comment]
Hi @franarolas 
Thanks for filing the issue

1.25 doesn't use canvaskit by default, it's HTML, canvaskit default change is recent and not present in 1.25
> @azbuky Always the default one, it should be always canvas as I am testing it through a desktop browser




### When using default on 1.25 (HTML) `flutter run -d chrome`
![Screenshot 2021-02-15 165712](https://user-images.githubusercontent.com/48603081/107941067-283d5480-6faf-11eb-9856-67b1442dca29.png)

### When forcing canvaskit on 1.25 (Canvaskit) ` flutter run --dart-define=FLUTTER_WEB_USE_SKIA=true -d chrome -v`
![Screenshot 2021-02-15 165327](https://user-images.githubusercontent.com/48603081/107941079-2f646280-6faf-11eb-9a05-1c3954a35a11.png)

Related https://github.com/flutter/flutter/issues/59082

Since you're getting performance benchmarks on 1.25, it must be HTML, can you please confirm? 
Otherwise, Can you please provide a complete reproducible minimal code sample to reproduce the performance report in 1.26 using canvaskit to be sure it's a regression
Thank you

**Comment by franarolas** [comment]
@TahaTesser 
The sample used was the default Flutter project.
If 1.25 was not using canvas by default then the test of the 1.25 is done with HTML instead of canvas, so it is not a regression.
Then the problem is in the canvaskit it self, the initial load time is way to high to be used in any production environment(I know Flutter web is in beta).

**Comment by yjbanov** [comment]
Applications using `--web-renderer=auto` and `--web-renderer=canvaskit` will take longer to start on empty browser cache. This is expected. We plan an effort to optimize it, but for now it's a deliberate trade-off. If your application requires the fastest possible start-up on empty browser cache, consider using `--web-renderer=html` instead.

However, if subsequent app reloads are slow in any mode, please let us know.

For now, I'm going to treat this issue as a feature request to make CanvasKit initial start-up time faster.

**Comment by franarolas** [comment]
@yjbanov Just testing the web apps that have been shown on the Flutter Engage today, the initial load with canvas is too slow to be in production, I think that you guys should put that on the top of the priorities.
Just check the following webs that you guys have shown today as examples:
[Flutter folio](https://www.flutterfolio.com/builds/latest/web-build-auto);
[Flutter plasma](https://flutterplasma.dev);
Once it is loaded the performance is amazing and if you have the website loaded in cache the initial load time is pretty good but from 0 to 100 it takes too much time.

**Comment by steeling** [comment]
The CDN hosting canvaskit is having issues.. can we enable an option to bundle canvaskit with our app, or in the interim simply move to a different CDN?

**Comment by ivanfeanor** [comment]
Same issue with flutter 2.2.0
Can be very easily reproduced by using fast 3g network speed in chrome - for me page load drops to 15-20s. With slow 3g it takes over a minute to load.

**Comment by cedvdb** [comment]
Also the js code is not tree shaken. I mean the library used and included inside the scripts in index.html. 

I believe this is something that has a high return on investment and can be implemented today by the flutter team (using existing tools ofc)

**Comment by BigGitWorld** [comment]
If the Flutter team can solve the initial load time problem (kavnvaskit 6MB),
I'm sure many people will use the Flutter to develop the web applications.
I have also written a web application with the Flutter and I am waiting for this problem to be solved soon.

**Comment by JMarji** [comment]
Does anyone know if there is a way to render a loading animation during the initial load?

**Comment by m0wer** [comment]
> Does anyone know if there is a way to render a loading animation during the initial load?

You can try https://stackoverflow.com/a/66535945.

**Comment by beersweetpicklepanda** [comment]
It's a single major issue for me in building flutter web,  otherwise I'm pretty happy of how things work in flutter.  Wish flutter 3.0 fixes this.

**Comment by qamaruddin-dev** [comment]
> Does anyone know if there is a way to render a loading animation during the initial load?

Follow this link 
https://itnext.io/fix-flutter-web-loading-time-with-a-loading-spinner-c5dd36a29f5b

**Comment by nicoste** [comment]
Hello everyone, after spending a lot of hours trying to solve this situation, I found a solution to speed up the first load of the app.
I saw that my `main.dart.js` was being downloaded and then, after that, the `main.dart.js` was triggering the download of the `canvaskit.wasm` file, this means that my app was starting on a cascade mode, avoiding downloading both files in parallel.

So my solution, after many many many attempts is based on adding only 2 lines of code to the `index.html` file (.../project_folder/web/index.html).
Inside the `<head>...</head>` tag insert this two lines:
```
<script src="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.js"></script>
<link rel="preload" href="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
```
I know hardcoding these URLs it's not a good practice, but, you will only need to change them if you update Flutter. I'm using Flutter `2.10.1` that uses `CanvasKit 0.30.0`, and when you build the web app, the generated `main.dart.js` has a hardcoded URL inside.

If you are going to add those 2 lines to your `index.html` file, I recommend you to build the web app, and open the `main.dart.js` file and search for "canvaskit-wasm", there you will find which is the URL you should use (something like "https://unpkg.com/canvaskit-wasm@0.33.0/bin/").

It's giving me a good improvement for the initial loading time! So, hope it works for you too!

**Comment by sawirricardo** [comment]
> Hello everyone, after spending a lot of hours trying to solve this situation, I found a solution to speed up the first load of the app. I saw that my `main.dart.js` was being downloaded and then, after that, the `main.dart.js` was triggering the download of the `canvaskit.wasm` file, this means that my app was starting on a cascade mode, avoiding downloading both files in parallel.
> 
> So my solution, after many many many attempts is based on adding only 2 lines of code to the `index.html` file (.../project_folder/web/index.html). Inside the `<head>...</head>` tag insert this two lines:
> 
> ```
> <script src="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.js"></script>
> <link rel="preload" href="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
> ```
> 
> I know hardcoding these URLs it's not a good practice, but, you will only need to change them if you update Flutter. I'm using Flutter `2.10.1` that uses `CanvasKit 0.30.0`, and when you build the web app, the generated `main.dart.js` has a hardcoded URL inside.
> 
> If you are going to add those 2 lines to your `index.html` file, I recommend you to build the web app, and open the `main.dart.js` file and search for "canvaskit-wasm", there you will find which is the URL you should use (something like "https://unpkg.com/canvaskit-wasm@0.33.0/bin/").
> 
> It's giving me a good improvement for the initial loading time! So, hope it works for you too!

Is this useful if we want to opt in to html renderer? and I think should we conditionally load this? say if html renderer, then dont download/include this.

**Comment by yahorsi** [comment]
Hi Guys, I just created an app using **"flutter create myapp"**, built it using **"flutter build web --release --web-renderer=html"** and it takes like 10 seconds or even more to run locally. Super slow for the super simple app that is actually starting sample app. Is that supposed to be like that or I'm doing something wrong? Can you please post here a sample of the published flutter web apps so I can compare performance? Additionally, there is an error in the console (could it be the reason for the slow startup?)

Basically what I want is just to build simplest app to test the speed, am I doing it wrong?

![image](https://user-images.githubusercontent.com/9819679/159078111-8164cb8a-5224-448a-a5e3-a8e726b829d5.png)

**Comment by cedvdb** [comment]
You built with the release flag but aren't you serving the debug app ? 

Side note the SEO score on lighthouse is misleading because the crawler can't crawl the content of the app.

**Comment by yahorsi** [comment]
@cedvdb no I serve to release, I did clean everything, then again created the app using "flutter create myapp" and then built using "flutter build web --release --web-renderer=htmд". I then created the web site in IIS and set the site folder to the ../build/web directory. When I open the site I see the following. 

![image](https://user-images.githubusercontent.com/9819679/159086093-ec70802d-af6c-4a5b-87ac-f2a00092d8b0.png)

**Comment by sawirricardo** [comment]
can we delete that `NOTICES` file?

**Comment by Zfinix** [comment]
> Hello everyone, after spending a lot of hours trying to solve this situation, I found a solution to speed up the first load of the app. I saw that my `main.dart.js` was being downloaded and then, after that, the `main.dart.js` was triggering the download of the `canvaskit.wasm` file, this means that my app was starting on a cascade mode, avoiding downloading both files in parallel.
> 
> So my solution, after many many many attempts is based on adding only 2 lines of code to the `index.html` file (.../project_folder/web/index.html). Inside the `<head>...</head>` tag insert this two lines:
> 
> ```
> <script src="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.js"></script>
> <link rel="preload" href="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
> ```
> 
> I know hardcoding these URLs it's not a good practice, but, you will only need to change them if you update Flutter. I'm using Flutter `2.10.1` that uses `CanvasKit 0.30.0`, and when you build the web app, the generated `main.dart.js` has a hardcoded URL inside.
> 
> If you are going to add those 2 lines to your `index.html` file, I recommend you to build the web app, and open the `main.dart.js` file and search for "canvaskit-wasm", there you will find which is the URL you should use (something like "https://unpkg.com/canvaskit-wasm@0.33.0/bin/").
> 
> It's giving me a good improvement for the initial loading time! So, hope it works for you too!

I just discovered `canvaskit-wasm` is bundled with the production webapp.
<img width="305" alt="image" src="https://user-images.githubusercontent.com/25801929/159104999-fdccc52e-430e-400f-9714-e80a4e81cc6b.png">

So by replacing `https://unpkg.com/canvaskit-wasm@0.33.0/bin/` with `your_url/` it forces it to load the already bundled one
<img width="641" alt="Screenshot 2022-03-19 at 4 31 01 AM" src="https://user-images.githubusercontent.com/25801929/159105048-80f4e9f5-de49-4216-85c0-1b3c5e90e45c.png">

**Comment by MurageKibicho** [comment]
> 

This is amazing! Forgive me for asking but how did you measure page download speed? It seems you're in dev tools but I don't know where this is.

**Comment by ditman** [comment]
We've been working on this lately. We just deployed to `master` a new bit of code to bootstrap Flutter Web apps from Javascript. One of the things that this allows is to create a loading animation that knows in what phase of the loading of Flutter the browser is (downloading main.dart.js, downloading additional engine resources, attempting to run the app...)

One example that will only let you start the flutter app when it's ready to do so:

* https://dit-bootstrap-tests.web.app

And the code we just pushed to `flutter/gallery` with the new initialization:

* https://github.com/flutter/gallery/blob/master/web/index.html#L77-L96

(See the demo [here, for now](https://dit-tests.web.app))

The new `flutter.js` API (`_flutter.loader.loadEntrypoint`...) will allow us all to experiment with more loading strategies like the ones described on this issue, and more! like @briansalvattore's chunking of the `main.dart.js`.

### But is it faster?

_Lighthouse_ seems to like the new approach better in my machine (compare dit-tests.web.app with gallery.flutter.dev before they deploy the new loading animation! 😉)

**Comment by sawirricardo** [comment]
do you have any steps to take for us to migrate from previous...version?

**Comment by ditman** [comment]
> do you have any steps to take for us to migrate from previous...version?

@sawirricardo for now, take a look at the code samples I provided above. You only need two things:

* flutter in channel master
* tweak your index.html, similarly to what was done [in this PR](https://github.com/flutter/gallery/pull/630) (you possibly have your own modifications, you'll want to keep those).

More thorough documentation is coming!

**Comment by jaumard** [comment]
@ditman interesting, I'm wondering if It allow a better error management? Because when an error occur before launching the app itself just nothing append, so you app stay in "loading" mode instead of showing a nice error messages. That's currently a pain to some of my users that just say "app is loading forever" as feedbacks. 
One of my user had a really bad internet connexion and his browser (firefox) wasn't able to load canvaskit.wasm at all, so for example in this case I would like to show an error message instead of doing nothing and app stay in loading mode.

**Comment by ditman** [comment]
@jaumard some things you could do in the new index.html:

1. Add a window.onerror listener that catches all runtime exceptions that bubble up to JavaScript, and attempt to send them to your backend, so you have a little bit of "error log" of unexpected exceptions. See [this](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror). That way you could see what is in the "JS Error console" of your users (if anything!)
2. You can create a promise "race" ([docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race)) with a timeout for your app (say: 10 seconds). If 10 seconds have elapsed, and the "engine Initializer" promise hasn't resolved (or the app, really whatever step you want), you can let people know that their connection is being "slow" and let them click-to-reload, or click to go to a separate (lighter) app. (Kind of what GMail does when it offers people to go "basic HTML").

**Comment by jaumard** [comment]
Interesting! thanks @ditman, I'll implement 1 to show an error message instead of "loading" if `flutter-first-frame` event is not fired before an error is raised. 
And I'll implement 2 this way:
default website uses canvas to render so I'll also generate an html renderer version an offer people to move to the html version if the custom timeout have been reached. 

And good news is that it can be done even with the stable web version :)

**Comment by jaumard** [comment]
Unfortunately the errors about canvas.wasm not being loaded doesn't go to windows.onerror for some reason (probably because it's of the main windows.
Here is simulate an issue by removing the file and getting 404. So I guess it will have to be caught by the race promise instead but it make users wait for nothing. 
<img width="547" alt="Capture d’écran 2022-04-20 à 22 22 07" src="https://user-images.githubusercontent.com/2970449/164316223-7da32728-9206-49ae-ad8d-fea9802bfcf9.png">

**Comment by ditman** [comment]
@jaumard We might be able to detect that case and make the load promise "fail fast" (as soon as canvaskit fails loading); for now the loading is too _optimistic_, and assumes that everything will be downloaded eventually.

I think I'll create a new issue for this use case. Internally we need to review what other *critical* resources fail to download, and how they communicate the failure to the loader.

**Comment by ditman** [comment]
@jaumard You should be able to use `window.onerror` to detect when canvaskit fails; but you must do it during the "capture" phase of the event (not the "bubbling" [phase](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Building_blocks/Events#event_bubbling_and_capture)). See this jsfiddle demo:

* https://jsfiddle.net/ditman/jcgf8h91

TL;DR:

```js
window.addEventListener('error', function(error) {
  console.log(error.target); // The tag that failed
  console.log('src:', error.target.src); // Check if 'src' contains 'canvaskit'?
}, {
  capture: true
});
```

* MDN: [Error event](https://developer.mozilla.org/en-US/docs/Web/API/Element/error_event)

**Comment by Kaminoan-Dev** [comment]
> > do you have any steps to take for us to migrate from previous...version?
> 
> @sawirricardo for now, take a look at the code samples I provided above. You only need two things:
> 
> * flutter in channel master
> * tweak your index.html, similarly to what was done [in this PR](https://github.com/flutter/gallery/pull/630) (you possibly have your own modifications, you'll want to keep those).
> 
> More thorough documentation is coming!

When using flutter in channel stable, is it "safe" just to copy and paste the `flutter.js` in the /web directory as a temporary fix while waiting for it to be merged to stable from master?

**Comment by ditman** [comment]
> When using flutter in channel stable, is it "safe" just to copy and paste the `flutter.js` in the /web directory as a temporary fix while waiting for it to be merged to stable from master?

@Kaminoan-Dev this will **not work with the current stable (`2.10.x`)**. The feature not only requires the `flutter.js` file, but also the Flutter Web framework knowing what to do with it from Dart :)

This feature will be available in the **next** stable after `2.10.x` (which is only a few days away).

**Comment by Kaminoan-Dev** [comment]
> > When using flutter in channel stable, is it "safe" just to copy and paste the `flutter.js` in the /web directory as a temporary fix while waiting for it to be merged to stable from master?
> 
> @Kaminoan-Dev this will **not work with the current stable (`2.10.x`)**. The feature not only requires the `flutter.js` file, but also the Flutter Web framework knowing what to do with it from Dart :)
> 
> This feature will be available in the **next** stable after `2.10.x` (which is only a few days away).

Understood. I will wait for the next release with great expectation :)
Thank you very much for you help and contributions!

**Comment by cedvdb** [comment]
it's possible to block the execution ?

The use case would be a landing page (HTML and css) eg `example.com`, with the app code downloaded in the background. Then when the user clicks on open app eg `example.app.com` the flutter code is already in the browser cache.

**Comment by ditman** [comment]
@cedvdb the new workflow only requires you to download a very small JS file that lets you control *when* the rest of the resources (main.dart.js, canvaskit.wasm, fonts...) get downloaded. Check the gallery.flutter.dev source (it's all in the index.html) Here's some [documentation](https://docs.flutter.dev/development/platform-integration/web/initialization).

**Comment by steeling** [comment]
the solution to this is to simply host canvaskit on a CDN that isn't abysmally slow (or allow us to specify our own CDN). The fact this hasn't been fixed is discouraging for the state of flutter web

I would love to hear one of the maintainer's chime in, but without an acknowledgment i don't know how web was designated as production ready

**Comment by muhleder** [comment]
@steeling Looks like you can already do that either with a dart define of FLUTTER_WEB_CANVASKIT_URL or by setting window.flutterConfiguration.canvasKitBaseUrl https://github.com/flutter/engine/blob/45bedbe84123a9ae0022fc95a6d0fc0f8ba9c10d/lib/web_ui/lib/src/engine/configuration.dart#L10

From memory it only works in release mode, and FLUTTER_WEB_CANVASKIT_URL didn't work due to a bug in dart which may be fixed now.

**Comment by steeling** [comment]
That's good to know about, thanks for pointing that out! @muhleder

**Comment by ashank96** [comment]
I have seen significant improvement in loading speed when the static resources like `main.dart.js` etc. is loaded as base64, in form of blob URIs. 
In order to reduce the loading time yet more - injecting `canvaskit.wasm` as base64 could help ? 
Currently unable to find a straightforward way to load canvaskit.wasm as base64 url inside `main.dart.js`. 

Any inputs on these lines would be appreciated. 
@ditman @yjbanov

**Comment by andreiborisov** [comment]
Another thing to consider: Safari currently doesn't cache `canvaskit.wasm` at all, loading it every time from the network🤦🏻‍♂️

**Comment by whiskeyPeak** [comment]
> I have seen significant improvement in loading speed when the static resources like `main.dart.js` etc. is loaded as base64, in form of blob URIs. 
> In order to reduce the loading time yet more - injecting `canvaskit.wasm` as base64 could help ? 
> Currently unable to find a straightforward way to load canvaskit.wasm as base64 url inside `main.dart.js`. 
> 
> Any inputs on these lines would be appreciated. 
> @ditman @yjbanov 

Could you provide some sample code? I'm really interested in using this approach!

**Comment by ugurberkecan** [comment]
> fast load flutter web flutter/website#6871

instead of decreasing the loading time, this increases my loading time.
I added this in the index.html file but also I couldn't see the canvas kit in main.dart.js

**Comment by JamshedAlamQaderi** [comment]
thank you @Zfinix. This helps me a lot.

**Comment by leithalnajjar** [comment]
any update?

**Comment by hillelcoren** [comment]
We've implemented this solution to load a local canvaskit file:

```
       <script>
         window.flutterConfiguration = {
           canvasKitBaseUrl: "https://example.com/my-custom-canvaskit/"
         };
       </script>
```

However we've run into a problem when updating our Flutter Web app. When there is a new version of the canvaskit file it looks like the old version is being cached (by CloudFlare and/or the browser) causing the app to fail to load. 

I think it may be helpful if the canvaskit link included the file version to act as a cache buster.

**Comment by TMSantos** [comment]
> We've implemented this solution to load a local canvaskit file:
> 
> ```
>        <script>
>          window.flutterConfiguration = {
>            canvasKitBaseUrl: "https://example.com/my-custom-canvaskit/"
>          };
>        </script>
> ```
> 
> However we've run into a problem when updating our Flutter Web app. When there is a new version of the canvaskit file it looks like the old version is being cached (by CloudFlare and/or the browser) causing the app to fail to load.
> 
> I think it may be helpful if the canvaskit link included the file version to act as a cache buster.

hey @hillelcoren , what load time improvements you experienced with this?

**Comment by ditman** [comment]
> I think it may be helpful if the canvaskit link included the file version to act as a cache buster.

@hillelcoren I've experienced this myself too, I'm going to create an issue so the canvaskit file that flutter vends contains its version.

Thanks for the feedback!

**Comment by JMarji** [comment]
What web server are you using?

With nginx you have to take some special configs for PWAs

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by JumperBot** [comment]
To those still relying on downloading `canvaskit.wasm` earlier and speading up the loading of your site instead of just doing [this](https://github.com/flutter/flutter/issues/76009#issuecomment-1072930101):

> Hello everyone, after spending a lot of hours trying to solve this situation, I found a solution to speed up the first load of the app. I saw that my `main.dart.js` was being downloaded and then, after that, the `main.dart.js` was triggering the download of the `canvaskit.wasm` file, this means that my app was starting on a cascade mode, avoiding downloading both files in parallel.
> 
> So my solution, after many many many attempts is based on adding only 2 lines of code to the `index.html` file (.../project_folder/web/index.html). Inside the `<head>...</head>` tag insert this two lines:
> 
> ```
> <script src="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.js"></script>
> <link rel="preload" href="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
> ```
> 
> I know hardcoding these URLs it's not a good practice, but, you will only need to change them if you update Flutter. I'm using Flutter `2.10.1` that uses `CanvasKit 0.30.0`, and when you build the web app, the generated `main.dart.js` has a hardcoded URL inside.
> 
> If you are going to add those 2 lines to your `index.html` file, I recommend you to build the web app, and open the `main.dart.js` file and search for "canvaskit-wasm", there you will find which is the URL you should use (something like "https://unpkg.com/canvaskit-wasm@0.33.0/bin/").
> 
> It's giving me a good improvement for the initial loading time! So, hope it works for you too!

It can be modified to work no matter what the latest version of canvaskit is by modifying it into:

```js
<script src="https://unpkg.com/canvaskit-wasm@latest/bin/canvaskit.js"></script>
<link rel="preload" href="https://unpkg.com/canvaskit-wasm@latest/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
```

Just a quick reminder:
There are better options than doing this to improve Flutter web app site load times!
An example of which is well... [this](https://github.com/flutter/flutter/issues/76009#issuecomment-1072930101)

**Comment by TMSantos** [comment]
> Hello everyone, after spending a lot of hours trying to solve this situation, I found a solution to speed up the first load of the app. I saw that my `main.dart.js` was being downloaded and then, after that, the `main.dart.js` was triggering the download of the `canvaskit.wasm` file, this means that my app was starting on a cascade mode, avoiding downloading both files in parallel.
> 
> So my solution, after many many many attempts is based on adding only 2 lines of code to the `index.html` file (.../project_folder/web/index.html). Inside the `<head>...</head>` tag insert this two lines:
> 
> ```
> <script src="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.js"></script>
> <link rel="preload" href="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
> ```
> 
> I know hardcoding these URLs it's not a good practice, but, you will only need to change them if you update Flutter. I'm using Flutter `2.10.1` that uses `CanvasKit 0.30.0`, and when you build the web app, the generated `main.dart.js` has a hardcoded URL inside.
> 
> If you are going to add those 2 lines to your `index.html` file, I recommend you to build the web app, and open the `main.dart.js` file and search for "canvaskit-wasm", there you will find which is the URL you should use (something like "https://unpkg.com/canvaskit-wasm@0.33.0/bin/").
> 
> It's giving me a good improvement for the initial loading time! So, hope it works for you too!

Doing this is making it worse for me, it downloads canvaskit twice

**Comment by JumperBot** [comment]
Has anybody found another exploit to make initial loading times faster?

**Comment by ditman** [comment]
@alangwutang11 just landed a commit in the Engine that will parallelize the download of canvaskit.wasm and the fonts that are required to start the app. That should make the app first load faster by about "whatever it took to load canvaskit.wasm before" (it removes one step of the download waterfall)

See this PR with lots of info:

https://github.com/flutter/engine/pull/36813

**Comment by zjrwdl** [comment]
> @alanwutang11 just landed a commit in the Engine that will parallelize the download of canvaskit.js and the fonts that are required to start the app.
> 
> That change should make the app first load faster by about "whatever it took to download the fonts after canvaskit" (it removes one step of the download waterfall)
> 
> See this PR with lots of info:
> 
> [flutter/engine#36813](https://github.com/flutter/engine/pull/36813)
> 
> _(Feature already available in the `master` channel, if you want to check when it becomes available in other tags, see [the commit](https://github.com/flutter/flutter/commit/14a736062211f9c85b4ba89de36d93f36d3a32ed))_

hello, i test latest version of flutter engine but it did not take effect.

Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.3.8, on macOS 11.6 20G165 darwin-x64, locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] Connected device (3 available)
[✓] HTTP Host Availability

<img width="1640" alt="image" src="https://user-images.githubusercontent.com/6874049/201003450-3cce4fb1-c2c7-40cf-a49d-2b3cc0b3429f.png">

**Comment by zjrwdl** [comment]
> the commit

<img width="484" alt="image" src="https://user-images.githubusercontent.com/6874049/201248149-c3e6e530-e4a7-42a5-aa34-0db053e94f9d.png">


<img width="577" alt="image" src="https://user-images.githubusercontent.com/6874049/201248035-edcea4f9-f062-46cd-8b94-2eaa1e9c19f8.png">


<img width="1572" alt="image" src="https://user-images.githubusercontent.com/6874049/201248015-46702796-2beb-42be-96ec-8be4d94e2904.png">

<img width="1003" alt="image" src="https://user-images.githubusercontent.com/6874049/201248273-c190aff0-dff9-46b0-98dd-0ebda8eabc2f.png">

**Comment by zjrwdl** [comment]
> @zjrwdl hey! try switching to the master channel using "flutter channel master" and see if the new change shows up?

<img width="484" alt="image" src="https://user-images.githubusercontent.com/6874049/201248149-c3e6e530-e4a7-42a5-aa34-0db053e94f9d.png">


<img width="577" alt="image" src="https://user-images.githubusercontent.com/6874049/201248035-edcea4f9-f062-46cd-8b94-2eaa1e9c19f8.png">


<img width="1572" alt="image" src="https://user-images.githubusercontent.com/6874049/201248015-46702796-2beb-42be-96ec-8be4d94e2904.png">

<img width="1003" alt="image" src="https://user-images.githubusercontent.com/6874049/201248273-c190aff0-dff9-46b0-98dd-0ebda8eabc2f.png">

@alanwutang11

**Comment by farr64** [comment]
Any idea when this speed fix for `canvaskit` may be released in the stable channel? Thanks.

**Comment by ditman** [comment]
@farr64 check this: https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#stable

**Comment by ditman** [comment]
@zjrwdl @alanwutang11 I've created a new issue so you can continue debugging/discussion there: https://github.com/flutter/flutter/issues/116223

**Comment by IdoBloch4** [comment]
About this comment: https://github.com/flutter/flutter/issues/76009#issuecomment-1072930101

> I just discovered canvaskit-wasm is bundled with the production webapp.
image

> So by replacing https://unpkg.com/canvaskit-wasm@0.33.0/bin/ with your_url/ it forces it to load the already bundled one
> Screenshot 2022-03-19 at 4 31 01 AM
> Screenshot 2022-03-19 at 4 31 59 AM

I've tried to replace it with my URL, but then it fetches the Canvaskit and the wasm twice.

<img width="709"  height="250" alt="Screen Shot 2022-12-28 at 1 18 13" src="https://user-images.githubusercontent.com/42539470/209735769-329a52bf-195e-4d32-9fb7-8727466e61ae.png">
<img width="722" height="250" alt="Screen Shot 2022-12-28 at 1 19 01" src="https://user-images.githubusercontent.com/42539470/209735764-b8dc85ed-8361-4fd4-a3ec-d37d2e501fcd.png">

Also, I see that my fonts are being downloaded after everything else is done, If that's ok I will add my example to #116223

**Comment by niloofaramerian** [comment]
> Hello everyone, after spending a lot of hours trying to solve this situation, I found a solution to speed up the first load of the app. I saw that my `main.dart.js` was being downloaded and then, after that, the `main.dart.js` was triggering the download of the `canvaskit.wasm` file, this means that my app was starting on a cascade mode, avoiding downloading both files in parallel.
> 
> So my solution, after many many many attempts is based on adding only 2 lines of code to the `index.html` file (.../project_folder/web/index.html). Inside the `<head>...</head>` tag insert this two lines:
> 
> ```
> <script src="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.js"></script>
> <link rel="preload" href="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
> ```
> 
> I know hardcoding these URLs it's not a good practice, but, you will only need to change them if you update Flutter. I'm using Flutter `2.10.1` that uses `CanvasKit 0.30.0`, and when you build the web app, the generated `main.dart.js` has a hardcoded URL inside.
> 
> If you are going to add those 2 lines to your `index.html` file, I recommend you to build the web app, and open the `main.dart.js` file and search for "canvaskit-wasm", there you will find which is the URL you should use (something like "https://unpkg.com/canvaskit-wasm@0.33.0/bin/").
> 
> It's giving me a good improvement for the initial loading time! So, hope it works for you too!

@nicoste  @Zfinix  I replaced my URL with  `https://unpkg.com/canvaskit-wasm@0.33.0/bin/`  but then it fetches the canvaskit.js and the canvaskit.wasm twice.

<img width="833" alt="Screenshot 1401-10-17 at 10 47 23" src="https://user-images.githubusercontent.com/69586214/211139420-854889f8-18a0-4c25-8d49-a47643142676.png">

**Comment by Samuel-ML** [comment]
I noticed that downloading the canvaskit files with the scripts:
```
<script src="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.js"></script>
<link rel="preload" href="https://unpkg.com/canvaskit-wasm@0.33.0/bin/canvaskit.wasm" as="fetch" crossOrigin="anonymous">
```
it doesn't work, it causes a double download and finally they are not used by `flutter_service_worker.js`

**Comment by ditman** [comment]
I would recommend to only use `link rel="preload"` if all that is needed is the browser to preload resources that are going to be used later, even for the `canvaskit.js` file, doing a [preload `as="script"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link#attr-as).

**Comment by yjbanov** [comment]
There are a few ongoing efforts to improve this:

- @hterkelsen is working on making a new custom build of CanvasKit that reduces the size by ~400kB.
- @mdebbar is working on reducing the size of the text layout sub-system by ~500kB.
- @ditman is working on various strategies for embedding Flutter Web in a web-page, that can help with load times as well.

We're going to treat this an the umbrella issue for improving load times, but a lot of work will be happening in smaller individual issues.

**Comment by DevinXiaoBai** [comment]
> There are a few ongoing efforts to improve this:
> 
> * @hterkelsen is working on making a new custom build of CanvasKit that reduces the size by ~400kB.
> * @mdebbar is working on reducing the size of the text layout sub-system by ~500kB.
> * @ditman is working on various strategies for embedding Flutter Web in a web-page, that can help with load times as well.
> 
> We're going to treat this an the umbrella issue for improving load times, but a lot of work will be happening in smaller individual issues.

Can you streamline the process of running canvaskit more，

**Comment by Ujjawalmaurya** [comment]
Deploy without canvaskit if its not much complex for rendering:
Build: `flutter build web --web-renderer=html`
then deploy

**Comment by almatov1** [comment]
Hello, @hterkelsen 
Any news about the canvaskit?

**Comment by harryfei** [comment]
We try to patch the generated `flutter.js` and `flutter_service_worker.js` to make flutter web first load faster. It's not perfect, but the result is acceptable to us. Wish it's helpful to you guys.

Those patches are based on Flutter master channel.

https://github.com/enkra/enkra-send/commit/a2ecf6b8f41e8ca35e8ceb62fb1a63fc3bf25c28

**Comment by crimsonvspurple** [comment]
> There are a few ongoing efforts to improve this:
> 
> * @hterkelsen is working on making a new custom build of CanvasKit that reduces the size by ~400kB.
> * @mdebbar is working on reducing the size of the text layout sub-system by ~500kB.
> * @ditman is working on various strategies for embedding Flutter Web in a web-page, that can help with load times as well.
> 
> We're going to treat this an the umbrella issue for improving load times, but a lot of work will be happening in smaller individual issues.

Hi.

Are you able to link issues for these listed things?

Does any of the strategy above includes breaking up canvaskit in parts so only the required parts are loaded first to accelerate first time flutter web visit?

**Comment by TMSantos** [comment]
@harryfei interesting, do you have any flutter website in production where I can test the improvements?

**Comment by harryfei** [comment]
@TMSantos Yeah, we use this patch in our app https://send.enkra.io.

Note that we add a loading indicator to avoid blank screen. So use the chrome dev tools to see real flutter first screen rendering time.

**Comment by tcakmak** [comment]
What do you think about this solution -> https://github.com/SchabanBo/deferred_test

**Comment by TMSantos** [comment]
@harryfei actually from everything that was said, your example is the only one that I can feel some improvement, well done

**Comment by commandiron** [comment]
> @TMSantos Yeah, we use this patch in our web app https://send.enkra.io
> 
> Note that we add a loading indicator to avoid blank screen. So use the chrome dev tools to see real flutter first screen rendering time.

How can i use this patch to my flutter web app? Can you explain step by step please?

**Comment by harryfei** [comment]
> > @TMSantos Yeah, we use this patch in our web app https://send.enkra.io
> > Note that we add a loading indicator to avoid blank screen. So use the chrome dev tools to see real flutter first screen rendering time.
> 
> How can i use this patch to my flutter web app? Can you explain step by step please?

1. Copy the files from this commit https://github.com/enkra/enkra-send/commit/a2ecf6b8f41e8ca35e8ceb62fb1a63fc3bf25c28 and replicate them in your Flutter repository.
2. Use flutter master channel `flutter channel master`
3. Execute `./build_tools/patch_web.sh` after every `flutter build web`.

**Comment by commandiron** [comment]
> > > @TMSantos Yeah, we use this patch in our web app https://send.enkra.io
> > > Note that we add a loading indicator to avoid blank screen. So use the chrome dev tools to see real flutter first screen rendering time.
> > 
> > 
> > How can i use this patch to my flutter web app? Can you explain step by step please?
> 
> 1. Copy the files from this commit [enkra/enkra-send@a2ecf6b](https://github.com/enkra/enkra-send/commit/a2ecf6b8f41e8ca35e8ceb62fb1a63fc3bf25c28) and replicate them in your Flutter repository.
> 2. Use flutter master channel `flutter channel master`
> 3. Execute `./build_tools/patch_web.sh` after every `flutter build web`.

Thanks for your help. flutter master channel is breaking various things, so I will have to wait with " fancy CircularProgressIndicator" XD

**Comment by razfazz** [comment]
> > > @TMSantos Yeah, we use this patch in our web app https://send.enkra.io
> > > Note that we add a loading indicator to avoid blank screen. So use the chrome dev tools to see real flutter first screen rendering time.
> > 
> > 
> > How can i use this patch to my flutter web app? Can you explain step by step please?
> 
> 1. Copy the files from this commit [enkra/enkra-send@a2ecf6b](https://github.com/enkra/enkra-send/commit/a2ecf6b8f41e8ca35e8ceb62fb1a63fc3bf25c28) and replicate them in your Flutter repository.
> 2. Use flutter master channel `flutter channel master`
> 3. Execute `./build_tools/patch_web.sh` after every `flutter build web`.

is the master channel mandatory to apply this patch?

**Comment by harryfei** [comment]
@razfazz Yeah. master channel has some modifications(fixing service worker cache bug) for `flutter_service_worker.js` which we need.

**Comment by jeancc29** [comment]
@harryfei Does it work on windows machine?

**Comment by harryfei** [comment]
@jeancc29 Yeah. it's a patch for flutter web which is just javascript and html. So It works on all OSes including Windows.

**Comment by eyebrowsoffire** [comment]
We've dropped the size of CanvasKit from 2.8MB compressed to 2.0MB compressed and 1.5MB compressed on Chromium-based browsers. We're going to close this issue for now, and track our further work on load time with follow-on issues.

**Comment by jaredsburrows** [comment]
@eyebrowsoffire Can you point us to the Git commit or Github PR? Also, what release of Flutter will have this update?

**Comment by eyebrowsoffire** [comment]
CanvasKit size changes are in Flutter 3.10. There is no single PR or git commit that is responsible for the size drop, it is a cumulative effect from many changes as part of a large effort to drop code size. The main things that contributed to the size drop:
* Lots of experimentation with build flags and inclusion/exclusion of features and code in CanvasKit
* Hosting CanvasKit on our own CDN and using brotli compression instead of gzip
* For Chromium-based browsers, we have a special build of CanvasKit that omits ICU data and leverages some specific Chrome APIs for text line breaking
